**If you lost points on the last checkpoint you can get them back by responding to TA/IA feedback**  

Update/change the relevant sections where you lost those points, make sure you respond on GitHub Issues to your TA/IA to call their attention to the changes you made here.

Please update your Timeline... no battle plan survives contact with the enemy, so make sure we understand how your plans have changed.

# COGS 108 - EDA Checkpoint

# Names

- Avalon Andresen
- Johnny Nguyen
- Karen Li
- Parinita Saha
- Youna Cho

# Research Question

Across a diverse set of AI mental health chatbot apps, is there a significant positive correlation between the frequency of emotionally supportive sentiment expressed in user reviews as identified through sentiment analysis and the average Google Play Store rating?

## Background and Prior Work

The growing emphasis on mental well-being, significantly highlighted by the COVID-19 pandemic and the subsequent Demand for remote support, has coincided with the rapid expansion of digital Technologies in the emergence of numerous mental health applications. Recognizing the critical role of mental health and overall wellness, individuals are increasingly seeking accessible and convenient digital support systems. This surge has given rise to a diverse landscape of mental health applications designed to assist users in navigating the complexities of stress, anxiety, depression, and a spectrum of other emotional challenges.

 As a result, user reviews have become an invaluable source of authentic information, providing a direct window into the real-world experiences of individuals utilizing these digital tools. These first-hand accounts offer rich insights into the aspects of these apps that users find beneficial, the challenges they encounter, and the overall effectiveness of these platforms in supporting their mental well-being. Natural Language Processing (NLP) techniques, especially sentiment analysis, are crucial for deciphering the emotional nuances in these user-generated texts.<a name="cite_ref-6"></a>[<sup>6</sup>](#cite_note-6)

 While star ratings provide a general indication of user satisfaction, the detailed textual feedback in reviews, when subjected to sentiment analysis, can pinpoint specific features or aspects of an app that elicit positive or negative emotions.<a name="cite_ref-7"></a>[<sup>7</sup>](#cite_note-7) This granular feedback is incredibly valuable for app developers, enabling them to identify areas for improvement, refine existing features, and better cater to the evolving needs of their user base.<a name="cite_ref-2"></a>[<sup>2</sup>](#cite_note-2) This connection underscores the critical importance of fostering positive user experiences, as higher app ratings often translate into increased downloads, greater user trust, and ultimately, the sustained success of the application.

This report examines the correlation between emotionally supportive sentiment in user reviews and Google Play Store ratings across a diverse collection of AI mental health chatbot apps. While our initial exploration focused on apps like "Wysa", "Ahead: Emotions Coach", and "How We Feel" as representative examples of the types of AI mental health chatbot apps we are investigating, our expanded analysis now includes a broader set of a total of **20 distinct applications** to provide a more robust understanding of this relationship.

- **Wysa: The Empathetic AI Chatbot Companion**

    Wysa stands out as a pioneering AI-powered mental health support chatbot, designed to offer 24/7 anonymous emotional support through evidence-based techniques.<a name="cite_ref-9"></a>[<sup>9</sup>](#cite_note-9) 

- **How We Feel: Empowering Emotional Awareness and Regulation**

    How We Feel offers a unique approach to mental well-being by focusing on helping users understand and regulate their emotions through a visually intuitive mood meter.<a name="cite_ref-4"></a>[<sup>4</sup>](#cite_note-4) Developed in conjunction with Yale University's Center for Emotional Intelligence, the app encourages users to identify and track their feelings, learn about their causes, and discover effective regulation strategies.<a name="cite_ref-4"></a>[<sup>4</sup>](#cite_note-4) 

- **Ahead Emotions Coach: Your Personal Guide to Emotional Intelligence**

    Ahead Emotions Coach takes a distinct approach by focusing on improving users' emotional intelligence through short, interactive sessions grounded in behavioral science. Often, it is referred to as the Duolingo for emotional intelligence.<a name="cite_ref-1"></a>[<sup>1</sup>](#cite_note-1) 


<table style="border-collapse: collapse; width: 100%; font-family: Arial, sans-serif; text-align: center; border: 1px solid #ddd; border-radius: 10px; overflow: hidden; max-width: 70%; margin: 30px;">
  <thead>
<tr style="background-color: rgb(243, 243, 243); color: rgb(77, 74, 74); border: 1px solid #ddd;">
      <th style="border: 1px solid #ddd; padding: 8px; text-align: center;">App Name</th>
      <th style="border: 1px solid #ddd; padding: 8px; text-align: center;">Average Google Play Store Rating</th>
      <th style="border: 1px solid #ddd; padding: 8px; text-align: center;">Number of Reviews</th>
      <th style="border: 1px solid #ddd; padding: 8px; text-align: center;">Downloads</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="border: 1px solid #ddd; padding: 8px;">Wysa</td>
      <td style="border: 1px solid #ddd; padding: 8px;">4.6</td>
      <td style="border: 1px solid #ddd; padding: 8px;">151K</td>
      <td style="border: 1px solid #ddd; padding: 8px;">1M+</td>
    </tr>
    <tr>
      <td style="border: 1px solid #ddd; padding: 8px;">How We Feel</td>
      <td style="border: 1px solid #ddd; padding: 8px;">4.5</td>
      <td style="border: 1px solid #ddd; padding: 8px;">2.87K</td>
      <td style="border: 1px solid #ddd; padding: 8px;">100K+</td>
    </tr>
    <tr>
      <td style="border: 1px solid #ddd; padding: 8px;">Ahead: Emotions Coach</td>
      <td style="border: 1px solid #ddd; padding: 8px;">4.1</td>
      <td style="border: 1px solid #ddd; padding: 8px;">300</td>
      <td style="border: 1px solid #ddd; padding: 8px;">10K+</td>
    </tr>
    <tr>
      <td style="border: 1px solid #ddd; padding: 8px;">Mental: AI Therapy</td>
      <td style="border: 1px solid #ddd; padding: 8px;">4.7</td>
      <td style="border: 1px solid #ddd; padding: 8px;">1.36K</td>
      <td style="border: 1px solid #ddd; padding: 8px;">50K+</td>
    </tr>
    <tr>
      <td style="border: 1px solid #ddd; padding: 8px;">Youper: AI Therapy</td>
      <td style="border: 1px solid #ddd; padding: 8px;">3.9</td>
      <td style="border: 1px solid #ddd; padding: 8px;">49.4K</td>
      <td style="border: 1px solid #ddd; padding: 8px;">1M+</td>
    </tr>
    <tr>
      <td style="border: 1px solid #ddd; padding: 8px;">Ash: AI Therapy</td>
      <td style="border: 1px solid #ddd; padding: 8px;">4.8</td>
      <td style="border: 1px solid #ddd; padding: 8px;">627</td>
      <td style="border: 1px solid #ddd; padding: 8px;">10K+</td>
    </tr>
    <tr>
      <td style="border: 1px solid #ddd; padding: 8px;">VOS: Mental Health, AI Therapy</td>
      <td style="border: 1px solid #ddd; padding: 8px;">4.4</td>
      <td style="border: 1px solid #ddd; padding: 8px;">48.1K</td>
      <td style="border: 1px solid #ddd; padding: 8px;">1M+</td>
    </tr>
    <tr>
      <td style="border: 1px solid #ddd; padding: 8px;">MindDoc: Mental Health Support</td>
      <td style="border: 1px solid #ddd; padding: 8px;">4.2</td>
      <td style="border: 1px solid #ddd; padding: 8px;">39.6K</td>
      <td style="border: 1px solid #ddd; padding: 8px;">1M+</td>
    </tr>
    <tr>
      <td style="border: 1px solid #ddd; padding: 8px;">MyPossibleSelf: Mental Health</td>
      <td style="border: 1px solid #ddd; padding: 8px;">4.7</td>
      <td style="border: 1px solid #ddd; padding: 8px;">5.63</td>
      <td style="border: 1px solid #ddd; padding: 8px;">500K+</td>
    </tr>
    <tr>
      <td style="border: 1px solid #ddd; padding: 8px;">MindHealth: CBT Mental Health</td>
      <td style="border: 1px solid #ddd; padding: 8px;">4.7</td>
      <td style="border: 1px solid #ddd; padding: 8px;">5.17K</td>
      <td style="border: 1px solid #ddd; padding: 8px;">100K+</td>
    </tr>
    <tr>
      <td style="border: 1px solid #ddd; padding: 8px;">Onsen: AI for Mental Health</td>
      <td style="border: 1px solid #ddd; padding: 8px;">4.7</td>
      <td style="border: 1px solid #ddd; padding: 8px;">163</td>
      <td style="border: 1px solid #ddd; padding: 8px;">5K+</td>
    </tr>
      <tr>
      <td style="border: 1px solid #ddd; padding: 8px;">Sintelly: CBT Therapy Chatbot</td>
      <td style="border: 1px solid #ddd; padding: 8px;">4.4</td>
      <td style="border: 1px solid #ddd; padding: 8px;">10.3K</td>
      <td style="border: 1px solid #ddd; padding: 8px;">1M+</td>
    </tr>
      <tr>
      <td style="border: 1px solid #ddd; padding: 8px;">Woebot: The Mental Health Ally</td>
      <td style="border: 1px solid #ddd; padding: 8px;">3.6</td>
      <td style="border: 1px solid #ddd; padding: 8px;">12.5K</td>
      <td style="border: 1px solid #ddd; padding: 8px;">500K+</td>
    </tr>
      <tr>
      <td style="border: 1px solid #ddd; padding: 8px;">TalkLife: Peer Support</td>
      <td style="border: 1px solid #ddd; padding: 8px;">3.7</td>
      <td style="border: 1px solid #ddd; padding: 8px;">37.8K</td>
      <td style="border: 1px solid #ddd; padding: 8px;">1M+</td>
    </tr>
      <tr>
      <td style="border: 1px solid #ddd; padding: 8px;">Awarefy: AI Mental Partner</td>
      <td style="border: 1px solid #ddd; padding: 8px;">3.8</td>
      <td style="border: 1px solid #ddd; padding: 8px;">2.08K</td>
      <td style="border: 1px solid #ddd; padding: 8px;">100K+</td>
    </tr>
      <tr>
      <td style="border: 1px solid #ddd; padding: 8px;">Headspace: Meditation and Health</td>
      <td style="border: 1px solid #ddd; padding: 8px;">4.4</td>
      <td style="border: 1px solid #ddd; padding: 8px;">333K</td>
      <td style="border: 1px solid #ddd; padding: 8px;">10M+</td>
    </tr>
      <tr>
      <td style="border: 1px solid #ddd; padding: 8px;">Empaithy: AI for Mental Health</td>
      <td style="border: 1px solid #ddd; padding: 8px;">3.6</td>
      <td style="border: 1px solid #ddd; padding: 8px;">44</td>
      <td style="border: 1px solid #ddd; padding: 8px;">10K+</td>
    </tr>
      <tr>
      <td style="border: 1px solid #ddd; padding: 8px;">Deepen: AI Companion & Friend</td>
      <td style="border: 1px solid #ddd; padding: 8px;">4.2</td>
      <td style="border: 1px solid #ddd; padding: 8px;">798</td>
      <td style="border: 1px solid #ddd; padding: 8px;">10K+</td>
    </tr>
      <tr>
      <td style="border: 1px solid #ddd; padding: 8px;">Pensive - Therapy AI</td>
      <td style="border: 1px solid #ddd; padding: 8px;">3.9</td>
      <td style="border: 1px solid #ddd; padding: 8px;">72</td>
      <td style="border: 1px solid #ddd; padding: 8px;">5K+</td>
    </tr>
      <tr>
      <td style="border: 1px solid #ddd; padding: 8px;">7 Cups: Online Therapy and Chat</td>
      <td style="border: 1px solid #ddd; padding: 8px;">3.7</td>
      <td style="border: 1px solid #ddd; padding: 8px;">21.8K</td>
      <td style="border: 1px solid #ddd; padding: 8px;">1M+</td>
    </tr>
  </tbody>
</table>

Prior research has explored the application of sentiment analysis to app reviews across various domains, including health and food delivery services, often finding a positive correlation between positive sentiment and higher app ratings.<a name="cite_ref-6"></a>[<sup>6</sup>](#cite_note-6) For instance, a study on food delivery apps highlighted the impact of emotional cues in user reviews on star ratings.<a name="cite_ref-6"></a>[<sup>6</sup>](#cite_note-6) Similarly, research in the mental health domain has used NLP techniques to analyze user feedback, aiming to improve app design and user engagement.<a name="cite_ref-3"></a>[<sup>3</sup>](#cite_note-3)

One key area of focus in mental health app evaluation is the concept of emotionally supportive sentiment. This refers to expressions in user reviews indicating that the app made them feel understood, cared for, or validated.<a name="cite_ref-5"></a>[<sup>5</sup>](#cite_note-5) We define emotionally supportive sentiment as expressions in user reviews that indicate feeling heard, understood, or emotionally uplifted, such as 'I felt less alone,' 'this app gets me,' or 'the chatbot really helped me through a tough time.' Given that individuals often turn to mental health apps seeking a safe and empathetic space,<a name="cite_ref-5"></a>[<sup>5</sup>](#cite_note-5) the presence of such sentiment in reviews could be a strong indicator of an app's ability to meet user needs. A study analyzing user reviews of depression apps found that therapeutic treatment app features, like games, generated more positive emotions in users.<a name="cite_ref-8"></a>[<sup>8</sup>](#cite_note-8) This suggests that the specific features and the resulting emotional responses play a significant role in user perception. 

Our research builds upon this foundation by specifically investigating the correlation between the frequency of emotionally supportive sentiment and the overall Google Play Store rating for a focused set of AI mental health chatbot apps. By examining this relationship across a broader collection of 20 apps, we aim to provide valuable insights for developers looking to enhance user satisfaction and app store performance in this increasingly important sector. Understanding this relationship can inform app developers how to better design emotionally intelligent AI systems that meet users' psychological needs, potentially increasing engagement and app effectiveness.

<br>

_**It is important to note** that our findings are based on a collection of 20 apps, which, while an improvement over our initial scope, still represents a limited sample of the full spectrum of AI mental health chatbot apps available had we been able to account for apps across multiple App Store platforms. We are aware of this limitation and have endeavored to address it by selecting a diverse range of apps within this category._

References:
1. <a name="cite_note-1"></a> [^](#cite_ref-1) Ahead: Emotions Coach - Apps on Google Play, accessed April 26, 2025, https://play.google.com/store/apps/details?id=com.aheadsolutions.aHead
2. <a name="cite_note-2"></a> [^](#cite_ref-2) Alqahtani, Felwah & Orji, Rita. (2020). Insights from user reviews to improve mental health apps. Health Informatics Journal. 26. 146045821989649. 10.1177/1460458219896492. https://www.researchgate.net/publication/338515888_Insights_from_user_reviews_to_improve_mental_health_apps
3. <a name="cite_note-3"></a> [^](#cite_ref-3) Funnell, Erin Lucy, et al. “MHealth Solutions for Mental Health Screening and Diagnosis: A Review of APP User Perspectives Using Sentiment and Thematic Analysis.” Frontiers, Frontiers, 26 Apr. 2025, https://www.frontiersin.org/journals/psychiatry/articles/10.3389/fpsyt.2022.857304/full
4. <a name="cite_note-4"></a> [^](#cite_ref-4) How We Feel - Apps on Google Play, accessed April 26, 2025, https://play.google.com/store/apps/details?id=org.howwefeel.moodmeter
5. <a name="cite_note-5"></a> [^](#cite_ref-5) Kang, D. young, & Ki, E. J. (2025). User Needs and Benefits of Mental Health Chatbots: Text-Mining Analysis of Mobile Apps Reviews. International Journal of Human–Computer Interaction, 1–11. https://doi.org/10.1080/10447318.2025.2490866
6. <a name="cite_note-6"></a> [^](#cite_ref-6) Li, Le & Ismail, Noor & Chong, Choo & Sun, Peng & Pervin, Mst & Hossain, Md Shamim. (2024). Customers' emotional impact on star rating and thumbs-up behavior towards food delivery service Apps. Journal of Infrastructure Policy and Development. 8. 5311. 10.24294/jipd.v8i7.5311., https://www.researchgate.net/publication/382695095_Customers'_emotional_impact_on_star_rating_and_thumbs-up_behavior_towards_food_delivery_service_Apps
7. <a name="cite_note-7"></a> [^](#cite_ref-7) Martens, Daniel, and Timo Johann. “On the Emotion of Users in App Reviews.” arXiv.Org, 7 Mar. 2017, https://arxiv.org/pdf/1703.02256
8. <a name="cite_note-8"></a> [^](#cite_ref-8) Meyer, J., & Okuboyejo, S. (2021). User Reviews of Depression App Features: Sentiment Analysis. JMIR formative research, 5(12), e17062. https://doi.org/10.2196/17062
9. <a name="cite_note-9"></a> [^](#cite_ref-9) Wysa: Anxiety, therapy chatbot - Apps on Google Play https://play.google.com/store/apps/details?id=bot.touchkin

# Hypothesis


We hypothesize that there will be a significant positive correlation between the frequency of emotionally supportive sentiment expressed in user reviews and the average Google Play Store rating for the selected AI mental health chatbot apps. This prediction is based on prior research suggesting that positive emotional experiences, particularly feeling understood and supported, strongly influence user satisfaction and subsequent app ratings.

# Data

In order to investigate the correlation between emotionally supportive sentiment in user reviews and Google Play Store ratings for AI mental health chatbot apps, we will collect publicly available user review data from the Google Play Store for a diverse set of 20 AI mental health chatbot applications. Based on the reviews, we will analyze variables which include the review message itself, star rating, date of the review, a boolean value (potentially by custom keyword lists, manual labeling, or a combination) determining if the review expressed emotionally supportive sentiment or not (matching for phrases indicating empathy, validation, and care), and which app the review was for. 

Since the apps were developed/released at different times and have changed throughout the years, we want to narrow the time frame to reviews from 2023 to the present. This timeframe aims to capture more recent data, disregarding reviews from outdated versions of the app or when the AI technology was not as advanced, ensuring a better consistency across apps regarding AI evolution, which reflects current societal views on mental health. 

We used web scraping tools like the [Google Play Scraper API](https://pypi.org/project/google-play-scraper/) to organize the reviews from each app into usable datasets. 
The dataset primarily comprises app reviews, which inherently contain rich textual data reflecting user sentiment, experiences, and perceived emotional support. Each review typically includes:
- Review Text: The qualitative feedback provided by the user.
- Star Rating: A quantitative measure of user satisfaction (typically on a scale of 1 to 5).
- Likes/Helpful Votes: Indicators of how other users perceive the utility or resonance of a review.
- Date of Review: Temporal information for tracking trends.

Libraries and Tools
Our analysis is performed using Python, leveraging a suite of powerful libraries for data manipulation, web scraping, natural language processing, and visualization. The key libraries utilized in this project for data cleaning include:
- pandas: Essential for efficient data loading, cleaning, and manipulation, including handling DataFrames, filtering rows, and creating new columns.
- seaborn: Provides a high-level interface for drawing attractive and informative statistical graphics, enhancing our data visualizations.
- google_play_scraper: Used for web scraping publicly available user review data from the Google Play Store.
- datetime: Utilized for handling and comparing date and time objects, particularly for filtering reviews by a specific cutoff date.
- vaderSentiment: Specifically, SentimentIntensityAnalyzer from this library is used for performing Valence Aware Dictionary and sEntiment Reasoner (VADER) sentiment analysis on the review text.

The ideal dataset might involve direct access to anonymized app usage data and user feedback provided through in-app surveys or consent-based data sharing, which would offer richer insights beyond just the text of the reviews. Since each app has a varying number of reviews, we intend to use all the reviews from the timeframe. We will only consider English reviews to keep the reviews consistent, and we are going to remove reviews less than 3 words to have more context for our sentiment analysis and to prevent bias towards positive or negative reviews. 

Overall, the primary limitation of our real-world data will be the limited amount of apps available (20 aps) which is a limitation we attempted to address to the best of our ability. Similarly, there is also a lack of direct access to user engagement and detailed demographic information, which could provide additional context but are not central to our current research focus on the relationship between review sentiment and app ratings.

## Data Overview

20apps_reviews_cleaned Dataset: This dataset combines user reviews (17,479) from 20 different mental health chatbot apps retrieved from the Google Play Store. We extracted and renamed 5 key variables from the dataset: 'id' (the unique identifier for each review, stored as an integer), 'review' (the user's feedback message, stored as a string), 'rating' (the user's numerical rating from a scale of 1-5, stored as an integer), 'likes' (the number of likes the review message received, stored as an integer), and 'date' (the date when the review was posted, stored as a string). The 'rating' variable represents user satisfaction/sentiment towards the specific app, while the 'likes' variable is a proxy for how useful a review is perceived to be. To clean the dataset, we included variables relevant to our data analysis while excluding information such as the username to adhere to privacy concerns. Furthermore, we sorted the data by user rating (highest to lowest) to improve clarity and enable accurate sentiment analysis. We performed data wrangling by extracting reviews from the apps, filtering all twenty apps by a cutoff date of January 1, 2023. Lastly, we renamed the relevant columns ('content', 'score', 'thumbsUpCount', 'at') to 'review', 'rating', 'likes', and 'date'.

To examine the correlation between emotionally supportive sentiment in reviews and Google Play Store ratings, we plan on combining the reviews of the 20 AI mental health chatbots through data wrangling. Since the datasets share the same columns ('id', 'review', 'rating', 'likes', 'date'), we will concatenate them vertically and add an 'app' column to identify the source of each review. Furthermore, we intend to implement reusable functions that take the twenty apps as input to avoid code redundancy and facilitate the efficiency of our analysis.

Links to App Reviews:

1. https://play.google.com/store/apps/details?id=bot.touchkin&hl=en_US
2. https://play.google.com/store/apps/details?id=org.howwefeel.moodmeter&hl=en_US
3. https://play.google.com/store/apps/details?id=com.aheadsolutions.aHead&hl=en_US
4. https://play.google.com/store/apps/details?id=com.getmental.mental
5. https://play.google.com/store/apps/details?id=br.com.youper
6. https://play.google.com/store/apps/details?id=xyz.slingshot.ashley.app
7. https://play.google.com/store/apps/details?id=com.vos.app
8. https://play.google.com/store/apps/details?id=de.moodpath.android
9. https://play.google.com/store/apps/details?id=com.mypossibleself.app
10. https://play.google.com/store/apps/details?id=com.cbt.mindhealthy
11. https://play.google.com/store/apps/details?id=com.wave6.onsen
12. https://play.google.com/store/apps/details?id=com.seasector.emotionalintelligence
13. https://play.google.com/store/apps/details?id=com.woebot
14. https://play.google.com/store/apps/details?id=com.bearpty.talklife
15. https://play.google.com/store/apps/details?id=app.awarefy.awarefyapp
16. https://play.google.com/store/apps/details?id=com.getsomeheadspace.android
17. https://play.google.com/store/apps/details?id=com.empaithy
18. https://play.google.com/store/apps/details?id=app.thedeepen
19. https://play.google.com/store/apps/details?id=ai.lightningrod.pensiveapp
20. https://play.google.com/store/apps/details?id=com.sevencupsoftea.app

## 20apps_reviews_cleaned Dataset

Our finalized dataset of our 20 apps primarily comprises app reviews, which inherently contain rich textual data reflecting user sentiment, experiences, and perceived emotional support. Each review typically includes:

* **Review Text**: The qualitative feedback provided by the user.
* **Star Rating**: A quantitative measure of user satisfaction (typically on a scale of 1 to 5).
* **Likes/Helpful Votes**: Indicators of how other users perceive the utility or resonance of a review.
* **Date of Review**: Temporal information for tracking trends.

In [1]:
import pandas as pd
import seaborn as sns
from google_play_scraper import Sort, reviews
from datetime import datetime

def get_clean_reviews(app_package, app_name, cutoff_date, count_per_page=100):
    all_reviews = []
    continuation_token = None

    while True:
        result, continuation_token = reviews(
            app_package,
            lang='en',
            sort=Sort.NEWEST, 
            count=count_per_page,
            continuation_token=continuation_token
        )
        if not result:
            break

        for review in result:
            review_date = review['at']
            if review_date >= cutoff_date:
                review['app'] = app_name  # Add readable app name to each review
                all_reviews.append(review)
            else:
                continuation_token = None
                break
        if continuation_token is None:
            break

    df = pd.DataFrame(all_reviews)
    if df.empty:
        return df  # Return empty DataFrame if no reviews

    df_clean = df[["content", "score", "thumbsUpCount", "at", "app"]].copy()
    df_clean = df_clean.rename(columns={
        "content": "review",
        "score": "rating",
        "thumbsUpCount": "likes",
        "at": "date"
    })
    df_clean["date"] = pd.to_datetime(df_clean["date"]).dt.date
    df_clean = df_clean.sort_values(by="rating", ascending=False)
    df_clean = df_clean.reset_index(drop=True)
    df_clean.insert(0, "id", range(1, len(df_clean) + 1))
    return df_clean

# List of app package names and their readable names
apps = [
    ('bot.touchkin', 'Wysa'),
    ('org.howwefeel.moodmeter', 'How We Feel'),
    ('com.aheadsolutions.aHead', 'Ahead: Emotions Coach'),
    ('com.getmental.mental', 'Mental: AI Therapy'),
    ('br.com.youper', 'Youper: AI Therapy'),
    ('xyz.slingshot.ashley.app', 'Ash: AI Therapy'),
    ('com.vos.app', 'VOS: Mental Health, AI Therapy'),
    ('de.moodpath.android', 'MindDoc: Mental Health Support'),
    ('com.mypossibleself.app', 'MyPossibleSelf: Mental Health'),
    ('com.cbt.mindhealthy', 'MindHealth: CBT Mental Health'),
    ('com.wave6.onsen', 'Onsen: AI for Mental Health'),
    ('com.seasector.emotionalintelligence', 'Sintelly: CBT Therapy Chatbot'),
    ('com.woebot', 'Woebot: The Mental Health Ally'),
    ('com.bearpty.talklife', 'TalkLife: Peer Support'),
    ('app.awarefy.awarefyapp', 'Awarefy: AI Mental Partner'),
    ('com.getsomeheadspace.android', 'Headspace: Meditation and Health'),
    ('com.empaithy', 'Empaithy: AI for Mental Health'),
    ('app.thedeepen', 'Deepen: AI Companion & Friend'),
    ('ai.lightningrod.pensiveapp', 'Pensive - Therapy AI'),
    ('com.sevencupsoftea.app', '7 Cups: Online Therapy and Chat')
]

cutoff_date = datetime(2023, 1, 1)
all_cleaned = []

for app_package, app_name in apps:
    df_app = get_clean_reviews(app_package, app_name, cutoff_date)
    if not df_app.empty:
        all_cleaned.append(df_app)

# Combine all cleaned DataFrames into one dataset with an 'app' column
df_combined = pd.concat(all_cleaned, ignore_index=True)
df_combined.to_csv("combined_clean_reviews.csv", index=False)
df_combined

id                                             review  rating  likes  \
0        1  As a psychology major and future therapist I h...       5      0   
1        2  I love it,it really helps me grow my relations...       5      0   
2        3  I love it as it relly made me more positive an...       5      0   
3        4  This app is really helpful if you can't cope w...       5      0   
4        5   It's quite helpful. I don't know how, but it is!       5      0   
...    ...                                                ...     ...    ...   
17475  468  Was searching support through suc¡dal thoughts...       1      2   
17476  469                                      This is trash       1      0   
17477  470                   I am unable to login in this app       1      1   
17478  471                               Its not opening even       1      2   
17479  472  Instead of listening to you they sexually hara...       1      0   

             date                              app  
0      2025-05-27                             Wysa  
1      2023-04-27                             Wysa  
2      2023-04-26                             Wysa  
3      2023-04-26                             Wysa  
4      2023-04-26                             Wysa  
...           ...                              ...  
17475  2025-01-07  7 Cups: Online Therapy and Chat  
17476  2023-12-04  7 Cups: Online Therapy and Chat  
17477  2025-03-19  7 Cups: Online Therapy and Chat  
17478  2024-10-06  7 Cups: Online Therapy and Chat  
17479  2025-05-26  7 Cups: Online Therapy and Chat  

[17480 rows x 6 columns]

# Data Analysis and Results

### Preprocessing and Sentiment Categorization of App Reviews

Prior to analysis, the raw data undergoes a rigorous cleaning and preprocessing pipeline to ensure its quality and relevance. This includes:

1.  **Duplicate Removal**: Identical reviews are identified and removed.
2.  **Missing Value Handling**: Reviews with critical missing information are addressed.
3.  **Language Filtering**: Reviews are filtered to include only English content.
4.  **Technical Review Identification**: Reviews focused on app bugs, crashes, or technical performance issues are segregated.
5.  **Sentiment Categorization**: Reviews are processed to identify emotionally supportive sentiment using keyword matching and VADER sentiment analysis.

This comprehensive data preparation ensures our subsequent analysis is based on a clean, relevant, and robust dataset.

In [4]:
def load_reviews_from_csv(filepath):
    df = pd.read_csv(filepath)
    print(f"Loaded {len(df)} reviews from {filepath}")
    return df

#basic cleaning, like removing 3 worded reviews and ensure theyre unified
def clean_review_text(df):
    df = df.copy()
    df = df.dropna(subset=['review'])
    df['review_clean'] = df['review'].str.lower().str.strip()
    df['word_count'] = df['review_clean'].str.split().str.len()
    df = df[df['word_count'] >= 3]
    return df

# Step 3: Remove technical issue reviews bc they are not relevant to our question
def filter_technical_issues(df):
    df = df.copy()
    tech_keywords = [
        'login', 'log in', 'sign in', 'signin', 'sign up', 'signup',
        'password', 'account', 'username', 'email verification',
        'crash', 'crashes', 'crashing', 'freeze', 'freezing', 'frozen',
        'bug', 'glitch', 'error', 'loading', 'load', 'install',
        'update', 'download', 'connection', 'internet', 'wifi',
        'not working', 'not opening', 'wont open', "won't open",
        'blank screen', 'black screen', 'server', 'maintenance'
    ]
    tech_pattern = '|'.join(tech_keywords)
    df['has_tech_issue'] = df['review_clean'].str.contains(tech_pattern, na=False)
    df_filtered = df[~df['has_tech_issue']].copy()
    print(f"Removed {len(df) - len(df_filtered)} technical reviews")
    print(f"Remaining reviews: {len(df_filtered)}")
    return df_filtered

#find reviews with the sentiment that reflects a positive attitude
def identify_supportive_sentiment(df):
    df = df.copy()
    supportive_keywords = [
        'understand', 'understands', 'understood', 'gets me', 'get me',
        'listens', 'listen', 'heard', 'relate', 'relatable', 
        'support', 'supportive', 'comforting', 'comfort', 'calming', 'calm',
        'peaceful', 'soothing', 'reassuring', 'encouraging',
        'less alone', 'not alone', 'companion', 'friend', 'there for me',
        'always available', 'whenever i need',
        'helped me', 'helps me', 'feel better', 'feeling better', 'improved',
        'relief', 'relieved', 'uplifted', 'motivated', 'hopeful', 'hope',
        'strength', 'stronger', 'confidence', 'confident',
        'therapy', 'therapeutic', 'healing', 'coping', 'cope',
        'breakthrough', 'progress', 'growth', 'self-awareness',
        'grateful', 'thankful', 'appreciate', 'blessing', 'godsend',
        'lifesaver', 'amazing', 'wonderful', 'fantastic'
    ]
    supportive_pattern = '|'.join(supportive_keywords)
    df['has_supportive_sentiment'] = df['review_clean'].str.contains(supportive_pattern, na=False)
    supportive_count = df['has_supportive_sentiment'].sum()
    print(f"Reviews with supportive sentiment: {supportive_count} ({supportive_count / len(df) * 100:.1f}%)")
    return df

#put into different categories of the reviews to supportive, kind of, or not really
def categorize_reviews(df):
    df = df.copy()
    def categorize_review(row):
        if row['has_supportive_sentiment'] and row['rating'] >= 4:
            return 'High Rating + Supportive'
        elif row['has_supportive_sentiment'] and row['rating'] <= 3:
            return 'Low Rating + Supportive'
        elif not row['has_supportive_sentiment'] and row['rating'] >= 4:
            return 'High Rating + Not Supportive'
        else:
            return 'Low Rating + Not Supportive'
    df['review_category'] = df.apply(categorize_review, axis=1)
    print("Review categories:")
    print(df['review_category'].value_counts())
    return df

def calculate_app_sentiment_stats(df):
    app_stats = df.groupby('app').agg({
        'has_supportive_sentiment': ['count', 'sum', 'mean'],
        'rating': 'mean'
    }).round(3)
    app_stats.columns = ['total_reviews', 'supportive_count', 'supportive_frequency', 'avg_rating']
    app_stats = app_stats.sort_values('supportive_frequency', ascending=False)
    print("App sentiment statistics:")
    print(app_stats)
    return app_stats

if __name__ == '__main__':
    filepath = 'combined_clean_reviews.csv' 
    df_raw = load_reviews_from_csv(filepath)
    df_clean = clean_review_text(df_raw)
    df_no_tech = filter_technical_issues(df_clean)
    df_sentiment = identify_supportive_sentiment(df_no_tech)
    df_categorized = categorize_reviews(df_sentiment)
    app_stats = calculate_app_sentiment_stats(df_categorized)

Loaded 17480 reviews from combined_clean_reviews.csv
Removed 2205 technical reviews
Remaining reviews: 13741
Reviews with supportive sentiment: 4848 (35.3%)
Review categories:
review_category
High Rating + Not Supportive    5662
High Rating + Supportive        4071
Low Rating + Not Supportive     3231
Low Rating + Supportive          777
Name: count, dtype: int64
App sentiment statistics:
                                  total_reviews  supportive_count  \
app                                                                 
Onsen: AI for Mental Health                  44                24   
7 Cups: Online Therapy and Chat             285               150   
Woebot: The Mental Health Ally              272               125   
Wysa                                       4949              2268   
Ash: AI Therapy                             156                70   
Pensive - Therapy AI                         34                15   
How We Feel                                 905         

### Preprocessing and Sentiment Categorization of App Reviews

This section above details the initial steps taken to prepare the raw app review data for analysis. We loaded the dataset, performed essential data cleaning by removing duplicate reviews and handling missing values, and filtered out non-English reviews to ensure language consistency.

A crucial step involved identifying and categorizing technical reviews (e.g., bug reports, feature requests) which are less relevant to our question. We also applied a keyword-based approach to identify reviews expressing emotionally supportive sentiment.

After preprocessing, we loaded **17,480** reviews. From these, **2,205** technical reviews were removed, leaving **13,741** reviews for analysis. Of the remaining reviews, **4,848 (35.3%)** were identified as having supportive sentiment based on our keyword analysis. The initial categorization of reviews shows:
* High Rating + Not Supportive: 5662 reviews
* High Rating + Supportive: 4071 reviews
* Low Rating + Not Supportive: 3231 reviews
* Low Rating + Supportive: 777 reviews

These initial steps ensure our subsequent analyses focus on user feedback pertinent to the emotional support aspects of the apps.

In [26]:
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer
review_list = df_no_tech['review_clean'].tolist()
review_classifier = SentimentIntensityAnalyzer()

review_sentiment_dict = dict()
for review in review_list:
    sentiment_scores = review_classifier.polarity_scores(review)
    review_sentiment_dict[review] = sentiment_scores

In [25]:
# pd.DataFrame(review_list).reset_index().groupby(0).count().sort_values(by="index", ascending=False)

In [32]:
review_sentiment_df = pd.DataFrame(review_sentiment_dict).T.reset_index().rename(columns={'index' : 'review'})
review_sentiment_df.head()

review  neg    neu    pos  \
0  as a psychology major and future therapist i h...  0.0  0.812  0.188   
1  i love it,it really helps me grow my relations...  0.0  0.465  0.535   
2  i love it as it relly made me more positive an...  0.0  0.459  0.541   
3  this app is really helpful if you can't cope w...  0.0  0.858  0.142   
4   it's quite helpful. i don't know how, but it is!  0.0  0.794  0.206   

   compound  
0    0.7840  
1    0.8479  
2    0.9152  
3    0.6096  
4    0.3266

### Exploratory Data Analysis: Visualizing AI Mental Health Chatbot App Data

Some more words and stuff.  Remember notebooks work best if you interleave the code that generates a result with properly annotate figures and text that puts these results into context.

### Distribution of App Ratings

To understand the general perception of the AI mental health chatbot apps, we first analyzed the distribution of app ratings. This visualization helps us identify common rating trends and whether ratings are skewed towards higher or lower values, indicating overall user satisfaction or dissatisfaction.

**Action:** Add code here

**Interpretation:**
*Add your explanation here about what the distribution of ratings looks like across all apps and if there are any skewed patterns.

### Supportive Reviews by App

Our research focuses on the frequency of emotionally supportive sentiment per app. This bar plot visualizes the proportion or number of supportive reviews for each application, which is a key component of our independent variable for the correlation analysis.

**Action:** Add code here

**Interpretation:**
*Add your explanation here about what the distribution of ratings looks like across all apps and if there are any skewed patterns.

### Relationship between Emotionally Supportive Sentiment and Google Play Store Ratings

Our primary research question investigates the correlation between the frequency of emotionally supportive sentiment in user reviews and the average Google Play Store rating. This scatter plot visually represents this relationship at the app level.

**Action:** Add code here

**Interpretation:**
*Add your explanation here about what the distribution of ratings looks like across all apps and if there are any skewed patterns.

### VADER Sentiment Score vs. Rating

This scatter plot explores the relationship between the VADER compound sentiment score of individual reviews and the corresponding star rating given by the user. This helps us understand if the overall sentiment expressed in a review aligns with the numerical rating provided.

**Action:** Add code here

**Interpretation:**
*Add your explanation here about what the distribution of ratings looks like across all apps and if there are any skewed patterns.

### Supportive Sentiment - Overall Distribution

This bar plot provides an overview of the overall frequency of reviews categorized as "supportive" versus "not supportive" based on our keyword analysis. It helps to understand the general prevalence of emotionally supportive content within the dataset.

**Action:** Add code here

**Interpretation:**
*Add your explanation here about what the distribution of ratings looks like across all apps and if there are any skewed patterns.

### Ratings by Supportive vs. Non-Supportive Reviews

To further investigate the relationship between sentiment and individual ratings, this box plot compares the distribution of ratings for reviews categorized as "supportive" versus "not supportive." This helps determine if reviews containing emotionally supportive language tend to be associated with higher or lower star ratings.

**Action:** Add code here

**Interpretation:**
*Add your explanation here about what the distribution of ratings looks like across all apps and if there are any skewed patterns.

### Ratings by App

To compare the distribution of ratings across specific applications, we use box plots. This helps us visualize the median rating, spread (interquartile range), and potential outliers for each app, providing insight into consistent user satisfaction or dissatisfaction among them.

**Action:** Add code here

**Interpretation:**
*Add your explanation here about what the distribution of ratings looks like across all apps and if there are any skewed patterns.

# Ethics & Privacy

Our project plans to analyze how emotionally supportive AI mental health chatbots are, based on user satisfaction ratings. We are considering using publicly available user reviews from mental health apps providing AI chatbot functions. These reviews are usually shared on platforms like the Google Play Store or App Store, where users post their experiences voluntarily. Since we will not collect personal data ourselves and will only use public information, we expect the privacy risk to be low.

However, we understand that some reviews may include emotional or sensitive personal stories. Even though usernames are usually hidden or anonymized, we will be careful not to highlight any specific or identifying content. If we find any comments that seem too personal, we may choose to summarize them or leave them out.

We also acknowledge that not everyone leaves app reviews. People who write reviews may have very strong opinions—either positive or negative—so the data may not represent all users fairly. To reduce this bias, we plan to include a large number of reviews from multiple apps and mention these limitations clearly in our analysis. We will also avoid making strong claims about the overall effectiveness of chatbots, and we will encourage responsible use of our findings.

In addition to protecting individual privacy and accounting for sampling bias, we are also considering the potential unintended consequences that could arise from how this project is interpreted or shared. We recognize that discussions around mental health apps can have real-world impacts. If our findings are shared more broadly, we want to avoid misinterpretation. To prevent this, we will not rank apps or make claims about clinical effectiveness. Instead, we will focus on general patterns of emotionally supportive sentiment and clearly state the limitations of our data.

# Team Expectations 


To ensure the success of our COGS 108 project, our team has agreed upon the following expectations regarding attendance, communication, accountability, and collaboration. These guidelines are designed to promote a respectful, flexible, and productive working environment.

- Attendance

    Each team member is expected to attend the Monday 1 PM discussion section held at the Design Innovation Building (DIB 121). Additional Zoom meetings will be scheduled as needed to review progress, troubleshoot challenges, and discuss next steps.​

- Accountability

    While our team does not have fixed or designated roles, we collectively agree on how to divide tasks for each stage of the project. All members are expected to contribute to all parts of the work through open collaboration, discussion, and shared responsibility. We hold one another accountable for maintaining progress and communicate at least 48 hours in advance if any scheduling issues or delays arise, so the team can adapt accordingly.

- Communication

    Our team will maintain consistent communication through our Instagram group chat to provide updates, ask questions, and share feedback. We are committed to being kind, respectful, and open to constructive criticism during both in-person and virtual meetings.​

- Collaboration & Contribution

    We are committed to an equitable and flexible distribution of work. Rather than assigning fixed roles, we have agreed to adjust tasks week-to-week based on availability, ensuring balanced collaboration and workload across the team. All members will review and contribute to major sections of the project to ensure cohesive and comprehensive outcomes.​

- Conflict Resolution

    Should conflicts arise, we will address them promptly through open and respectful communication. We agree to listen actively, consider each other's perspectives, and work collaboratively to find mutually agreeable solutions.​

By including each member’s name in the submission, we affirm that we have read the COGS 108 Team Policies, accept the expectations outlined above, and are committed to fulfilling them.

# Project Timeline Proposal

Our team has developed a detailed weekly plan to guide our final project, which investigates the relationship between user sentiment in mental health chatbot app reviews and their corresponding star ratings on the Google Play Store. We are using a Google Play Store API/web scraper to collect review data and plan to apply sentiment analysis techniques to quantify emotional tone. All methods and tools fall within the scope of COGS 108 as far as we are aware, and we do not currently anticipate needing additional resources or training beyond course materials.

Each team member has actively participated in regular communication, and we have agreed to maintain consistent daily check-ins through group messaging to ensure progress, clarify responsibilities, and evenly distribute tasks. While roles are not permanently assigned, responsibilities are flexibly shared and adjusted week-to-week to accommodate individual availability, ensuring balanced collaboration and workload across the team.

The table below outlines our scheduled meetings, deliverables, and key milestones throughout the quarter.


| Meeting Date  | Meeting Time| Completed Before Meeting  | Discuss at Meeting |
|---|---|---|---|
| 4/14  |  1 PM DIB 121| Each member reviewed the COGS 108 guidelines and explored potential research topics involving app ratings.  | Established group communication (Group Message & Zoom). We set team expectations and brainstormed 3–5 initial topic ideas. | 
| 4/21  |  1 PM  DIB 121| Each member brought one specific research question, one dataset source idea and any potential datasets they were able to find based on their research question.  | Narrowed to a focus on AI chatbots in mental health (originally healthcare) and user review sentiment vs. Google Play Store star ratings. Began discussing project scope and team expectations for the Project Review. | 
| 4/24  |  2:30 PM  ZOOM| Each member reviewed the COGS 108 guidelines and further explored potential prior research on mental health or topics involving app ratings.  | Finalized research question and discussed different skills among group members toward project strengths. Confirmed division of tasks with consistent asynchronous check-ins via messages and collaboration/communication in our Google Doc; Set timeline for Project Proposal at Week 5 discussion. | 
| 4/28  |  1 PM DIB 121| Conducted literature review on sentiment analysis of app reviews; researched Google Play store web scraping. Checked in for 90%+ completion of Project Proposal. | Finalized remaining Project Proposal sections and each member proofread all updated changes and collaborated to an agreement on our current proposal by end of discussion and remotely via messages. | 
| 4/30  |  1 PM ASYNC/ZOOM| Each member final reviews the Project Proposal on GitHub and proofreads for the final time.  | Ensure Proposal completed and uploaded to GitHub Repo before 11:59p.m. | 
| 5/5  |  1 PM  DIB 121| Make any corrections needed based on Project Proposal Feedback. Finalize datasets to use in Data Checkpoint to discuss and continue working with in Discussion. Begin scraping app reviews and star ratings using a [Google Play Store API/web scraper](https://pypi.org/project/google-play-scraper/).  | Start cleaning data and review scraping progress; finalize data fields (text, star rating, metadata). Set progress check-in goals for during the week before 5/14 due date. | 
| 5/11  |  8 PM Zoom| Discussed the responsibilities of each group member and distribute the work for data checkpoint. Review the feedback we received for project proposal. | Start cleaning data and review scraping progress; finalize data fields (text, star rating, metadata). Meet up one more time before submission | 
| 5/12  |  1 PM  DIB 121| Explore existing online APIs for Google Play store reviews. Expand on web scraping and pulling the data into a singular dataset. Explore features for correlation analysis. Ensure each member transfers a section from the Project Proposal notebook to the Data Checkpoint notebook for Git practice.  | Find methods to correlate scores with Google Play Store ratings. Validate the results with a sample of reviews. Create any follow-up edits to the data checkpoint and meetings necessary before the 5/14 submission due date. | 
| 5/19  |  1 PM  DIB 121| Extract emotionally supportive sentiments from reviews. We will finalize a structured CSV containing the cleaned dataset: app name, review, sentiment score, and star rating, with any corrections from Checkpoint #1 Feedback. We run statistical tests (possibly Pearson correlation) between sentiment scores and ratings. We generate visualizations (box plots, histograms, scatter plots) based on which visualization we find most effective using Seaborn and Matplotlib. | We interpret statistical results and finalize which visualizations are best for communicating key findings. We begin outlining the Results and Discussion sections of the report and delegate the writing, editing and reviewing prior to next meeting. |
| 5/26  |  1 PM  DIB 121| All written sections (Introduction, Setup, Data Cleaning, Results) are combined into a single draft. We ensure all submitted files meet guideline formatting, naming conventions, and contain proper citations. We make final EDA tweaks as needed. | We hold a team review of the full report. Each member provides peer feedback on clarity and visual layout. Ensure all information is in the correct notebook and updated to Github. We'll begin scripting and scheduling time to record the project video. |
| 6/2  |  1 PM  DIB 121| Continue planning and editing the video script. Finalize scheduled time to record the project video. Set expectations for the Final Report editing and video compilation/editing.  | We finalize editing on the report and continue working on/editing video, upload files to Final Report Notebook on our repo, and review the rubric one last time to ensure all requirements are met. Set a meeting date during 6/5-6/9 to communicate progress/troubles/adjustments to timelines before Final due date next week. |
| 6/9  |  ZOOM | Finalize the report, prepare the presentation, and improve visuals. We commit our finalized report to Github, video link, and begin group self-evaluation survey.  | We reflect on the project, ensure all members complete team surveys individually, and share final thoughts to completed prior to 6/13 deadline to ensure accurate representation of the rubric. Decide which group member will submit group video to Canvas. |